### PR TITLE
fixed agency bug, when started from compaction

### PR DESCRIPTION
**Pull request by kvahed**

### CODE REVIEW

File: arangod/Agency/Agent.cpp [hunk 1]
```diff
@@ -396,9 +398,11 @@ priv_rpc_ret_t Agent::recvAppendEntriesRPC(
 
 /// Leader's append entries
 void Agent::sendAppendEntriesRPC() {
+  LOG_TOPIC(INFO, Logger::AGENCY) << __FILE__<< __LINE__;
```
Comment (neunhoef) [inline arangod/Agency/Agent.cpp]: This looks like debugging output which we should remove or set the log level to debug.

File: arangod/Agency/Agent.cpp [hunk 2]
```diff
@@ -396,9 +398,11 @@ priv_rpc_ret_t Agent::recvAppendEntriesRPC(
...
   auto cc = ClusterComm::instance();
   if (cc == nullptr) {
     // nullptr only happens during controlled shutdown
+    LOG_TOPIC(INFO, Logger::AGENCY) << __FILE__<< __LINE__;
```
Comment (neunhoef) [inline arangod/Agency/Agent.cpp]: Same here.

File: arangod/Agency/Agent.cpp [hunk 3]
```diff
@@ -440,8 +444,14 @@ void Agent::sendAppendEntriesRPC() {
         READ_LOCKER(oLocker, _outputLock);
         commitIndex = _commitIndex;
       }
-      std::vector<log_t> unconfirmed = _state.get(lastConfirmed, lastConfirmed+99);
 
+      if (lastConfirmed < _state.firstIndex) {
```
Comment (neunhoef) [inline arangod/Agency/Agent.cpp]: It seems that this does not compile, since firstIndex is a method.

File: arangod/Agency/Agent.cpp [hunk 4]
```diff
@@ -348,18 +350,41 @@ priv_rpc_ret_t Agent::recvAppendEntriesRPC(
     return priv_rpc_ret_t(false,t);
   }
 
+  size_t nqs = payload.length();
+  if(nqs > 0 && !payload[0].get("readDB").isNone()) {
+    // We have received a compacted state.
+    // Whatever we got in out own state is meaningless now. It is a new world.
```
Comment (neunhoef) [inline arangod/Agency/Agent.cpp]: "out" -> "our"

File: arangod/Agency/Agent.cpp [hunk 5]
```diff
@@ -348,18 +350,41 @@ priv_rpc_ret_t Agent::recvAppendEntriesRPC(
     return priv_rpc_ret_t(false,t);
   }
 
+  size_t nqs = payload.length();
+  if(nqs > 0 && !payload[0].get("readDB").isNone()) {
```
Comment (neunhoef) [inline arangod/Agency/Agent.cpp]: Is this clever? Assume we are at index 2000 but our last snapshot is at 1000. If we now get an appendEntriesRPC with a snapshot at 1500 and 99 more log entries, should we then not check if our log fits to the snapshot and data and not change anything if everything is OK? With this, we are throwing away a huge performance improvement which we have already implemented and which works. I think we must be more clever here and perform the check even when we receive a snapshot. Remember that snapshotting could be done differently in leader and follower!

File: arangod/Agency/Agent.cpp [hunk 6]
```diff
@@ -451,14 +488,13 @@ void Agent::sendAppendEntriesRPC() {
...
+      if (unconfirmed.empty() && !raiseLastConfirmed) {
         LOG_TOPIC(ERR, Logger::AGENCY) << "Unexpected empty unconfirmed: "
           << "lastConfirmed=" << lastConfirmed << " commitIndex="
           << commitIndex;
+        TRI_ASSERT(false);
       }
 
-      TRI_ASSERT(!unconfirmed.empty());
-
       if (unconfirmed.size() == 1) {
```
Comment (neunhoef) [inline arangod/Agency/Agent.cpp]: Is this still correct with `raiseLastConfirmed`? I think if `raiseLastConfirmed` is `true`, we should not handle the size 1 case in this way.

File: arangod/Agency/Agent.cpp [hunk 7]
```diff
@@ -440,8 +465,20 @@ void Agent::sendAppendEntriesRPC() {
         READ_LOCKER(oLocker, _outputLock);
         commitIndex = _commitIndex;
       }
-      std::vector<log_t> unconfirmed = _state.get(lastConfirmed, lastConfirmed+99);
 
+      // If lastConfirmed is smaller than our first log entry's index, and
+      // given that our first log entry is either the 0-entry or a compacted
```
[the remark this answers is not in the document]
Comment (kvahed) [reply arangod/Agency/Agent.cpp]: Corrected

File: arangod/Agency/Agent.cpp (hunk 6 above)
Comment (kvahed) [reply arangod/Agency/Agent.cpp]: fixed. manually testing

File: arangod/Agency/Agent.cpp (hunk 5 above)
Comment (kvahed) [reply arangod/Agency/Agent.cpp]: will leave as is for now.

**Comment by neunhoef** [comment]
Superseded by other PR #5828 which is now merged

**Pull request closed without merging**